File: apps/starter-ios/StarterApp/Core/Subscription/SubscriptionManager.swift
import Foundation
import StoreKit

enum SubscriptionTier: String {
    case free = "free"
    case tracker = "tracker"
}

enum TrackerFeature {
    case basicLogging, exerciseLibrary             // Always free
    case unlimitedTemplates, advancedAnalytics, exportData  // Tracker+
}

@Observable
@MainActor
final class SubscriptionManager {
    static let shared = SubscriptionManager()

    var currentTier: SubscriptionTier = .free
    var isLoading: Bool = false

    // Product IDs — must match App Store Connect exactly
    private let trackerMonthlyId = "tracker_monthly"
    private let trackerYearlyId  = "tracker_yearly"
    private let trackerLifetimeId = "tracker_lifetime"

    private var products: [Product] = []
    nonisolated(unsafe) private var purchaseListener: Task<Void, Never>?

    init() {
        purchaseListener = listenForTransactions()
        Task { await loadProducts() }
    }

    deinit {
        purchaseListener?.cancel()
    }

    // MARK: — Products

    func loadProducts() async {
        isLoading = true
        do {
            products = try await Product.products(for: [
                trackerMonthlyId, trackerYearlyId, trackerLifetimeId,
            ])
        } catch {
            // Silent failure — paywall shows empty state with retry
        }
        isLoading = false
    }

    func product(for id: String) -> Product? {
        products.first { $0.id == id }
    }

    // MARK: — Purchase

    func purchase(_ product: Product) async throws {
        let result = try await product.purchase()
        switch result {
        case .success(let verification):
            let transaction = try checkVerified(verification)
            await transaction.finish()
            await refreshEntitlements()
        case .userCancelled:
            break
        case .pending:
            break
        @unknown default:
            break
        }
    }

    func restorePurchases() async throws {
        try await AppStore.sync()
        await refreshEntitlements()
    }

    // MARK: — Entitlements

    func refreshEntitlements() async {
        for await result in Transaction.currentEntitlements {
            guard let transaction = try? checkVerified(result) else { continue }
            if transaction.productID == trackerMonthlyId ||
               transaction.productID == trackerYearlyId ||
               transaction.productID == trackerLifetimeId {
                currentTier = .tracker
                return
            }
        }
        currentTier = .free
    }

    func resetEntitlements() {
        currentTier = .free
    }

    func isFeatureAvailable(_ feature: TrackerFeature) -> Bool {
        switch feature {
        case .basicLogging, .exerciseLibrary:
            return true
        case .unlimitedTemplates, .advancedAnalytics, .exportData:
            return currentTier == .tracker
        }
    }

    // MARK: — Transaction listener

    private func listenForTransactions() -> Task<Void, Never> {
        Task.detached {
            for await result in Transaction.updates {
                guard let transaction = try? self.checkVerified(result) else { continue }
                await self.refreshEntitlements()
                await transaction.finish()
            }
        }
    }

    nonisolated private func checkVerified<T>(_ result: VerificationResult<T>) throws -> T {
        switch result {
        case .unverified: throw StoreError.failedVerification
        case .verified(let value): return value
        }
    }
}

enum StoreError: Error {
    case failedVerification
}

File: apps/starter-android/app/src/main/kotlin/com/onlystack/starterapp/core/subscription/SubscriptionManager.kt
package com.onlystack.starterapp.core.subscription

import android.content.Context
import com.android.billingclient.api.*
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

enum class SubscriptionTier { FREE, TRACKER }

enum class TrackerFeature {
    BASIC_LOGGING, EXERCISE_LIBRARY,                           // Free
    UNLIMITED_TEMPLATES, ADVANCED_ANALYTICS, EXPORT_DATA,     // Tracker+
}

@Singleton
class SubscriptionManager @Inject constructor(
    @ApplicationContext private val context: Context,
) : PurchasesUpdatedListener {

    // Product IDs — must match Play Console exactly
    private val trackerMonthlyId  = "tracker_monthly"
    private val trackerYearlyId   = "tracker_yearly"
    private val trackerLifetimeId = "tracker_lifetime"

    private val _currentTier = MutableStateFlow(SubscriptionTier.FREE)
    val currentTier: StateFlow<SubscriptionTier> = _currentTier.asStateFlow()

    private val scope = CoroutineScope(Dispatchers.IO)

    private val billingClient: BillingClient = BillingClient.newBuilder(context)
        .setListener(this)
        .enablePendingPurchases(
            PendingPurchasesParams.newBuilder().enableOneTimeProducts().build()
        )
        .build()

    init {
        connect()
    }

    // MARK: — Connection

    private fun connect() {
        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(result: BillingResult) {
                if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                    scope.launch { refreshEntitlements() }
                }
            }
            override fun onBillingServiceDisconnected() {
                connect() // auto-reconnect
            }
        })
    }

    // MARK: — Products

    suspend fun queryProducts(): List<ProductDetails> {
        val productList = listOf(trackerMonthlyId, trackerYearlyId).map { id ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(id)
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        } + listOf(
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(trackerLifetimeId)
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
        )

        val params = QueryProductDetailsParams.newBuilder()
            .setProductList(productList)
            .build()

        return suspendCancellableCoroutine { cont ->
            billingClient.queryProductDetailsAsync(params) { _, details ->
                cont.resume(details)
            }
        }
    }

    // MARK: — Purchase

    fun launchPurchaseFlow(activity: android.app.Activity, productDetails: ProductDetails) {
        val offerToken = productDetails.subscriptionOfferDetails?.firstOrNull()?.offerToken
        val productDetailsParamsList = if (offerToken != null) {
            listOf(
                BillingFlowParams.ProductDetailsParams.newBuilder()
                    .setProductDetails(productDetails)
                    .setOfferToken(offerToken)
                    .build()
            )
        } else {
            listOf(
                BillingFlowParams.ProductDetailsParams.newBuilder()
                    .setProductDetails(productDetails)
                    .build()
            )
        }
        val flowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(productDetailsParamsList)
            .build()
        billingClient.launchBillingFlow(activity, flowParams)
    }

    override fun onPurchasesUpdated(result: BillingResult, purchases: List<Purchase>?) {
        if (result.responseCode == BillingClient.BillingResponseCode.OK && purchases != null) {
            scope.launch {
                purchases.forEach { handlePurchase(it) }
            }
        }
    }

    private suspend fun handlePurchase(purchase: Purchase) {
        if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
            if (!purchase.isAcknowledged) {
                val params = AcknowledgePurchaseParams.newBuilder()
                    .setPurchaseToken(purchase.purchaseToken)
                    .build()
                suspendCancellableCoroutine { cont ->
                    billingClient.acknowledgePurchase(params) { cont.resume(it) }
                }
            }
            refreshEntitlements()
        }
    }

    // MARK: — Entitlements

    suspend fun refreshEntitlements() {
        val subsResult = billingClient.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        )
        val inappResult = billingClient.queryPurchasesAsync(
            QueryPurchasesParams.newBuilder()
                .setProductType(BillingClient.ProductType.INAPP)
                .build()
        )
        val allPurchases = subsResult.purchasesList + inappResult.purchasesList
        val hasTracker = allPurchases.any { purchase ->
            purchase.purchaseState == Purchase.PurchaseState.PURCHASED &&
            purchase.products.any { it in listOf(trackerMonthlyId, trackerYearlyId, trackerLifetimeId) }
        }
        _currentTier.value = if (hasTracker) SubscriptionTier.TRACKER else SubscriptionTier.FREE
    }

    fun resetEntitlements() {
        _currentTier.value = SubscriptionTier.FREE
    }

    fun isFeatureAvailable(feature: TrackerFeature): Boolean = when (feature) {
        TrackerFeature.BASIC_LOGGING, TrackerFeature.EXERCISE_LIBRARY -> true
        TrackerFeature.UNLIMITED_TEMPLATES,
        TrackerFeature.ADVANCED_ANALYTICS,
        TrackerFeature.EXPORT_DATA -> _currentTier.value == SubscriptionTier.TRACKER
    }
}
